sklearn/metrics/classification.py

```diff
@@ -268,8 +268,14 @@ def confusion_matrix(y_true, y_pred, labels=None, sample_weight=None):
     # also eliminate weights of eliminated items
     sample_weight = sample_weight[ind]
 
+    # Choose the accumulator dtype to always have high precision
+    if sample_weight.dtype.kind in {'i', 'u', 'b'}:
+        dtype = np.int64
+    else:
+        dtype = np.float64
```

Comment:
I don't understand the logic of upcasting everything to the maximum resolution.

Typically, I expect code to keep the same types as what I put in. If I put in float32, it is often a choice, to limit memory consumption.